src/pie_utils/sequence_tagging/encoding.py

```diff
@@ -0,0 +1,363 @@
+    base_sequence_length: int,
+    offset: int = 0,
+    include_ill_formed: bool = False,
+) -> List[str]:
+    # create IOB2 encoding
+    tags = ["O"] * base_sequence_length
+    labeled_spans = sorted(labeled_spans, key=lambda span_annot: span_annot[1][0])
+    for i, (label, (start, end)) in enumerate(labeled_spans):
+        _start = start - offset
+        _end = end - offset
+
+        previous_tags = tags[_start:_end]
+        if previous_tags != ["O"] * len(previous_tags):
+            # raise ValueError(f"tags already set [{previous_tags}], i.e. there is an annotation overlap")
+            if not include_ill_formed:
+                continue
+
+        # create IOB2 encoding
+        tags[_start] = f"B-{label}"
+        tags[_start + 1 : _end] = [f"I-{label}"] * (len(previous_tags) - 1)
+
+    return tags
+
+
+def token_spans_to_tag_sequence(
+    labeled_spans: List[TypedStringSpan],
+    base_sequence_length: int,
+    coding_scheme: str = "IOB2",
+    offset: int = 0,
+    include_ill_formed: bool = True,
+) -> List[str]:
+
+    tags = labeled_spans_to_iob2(
+        labeled_spans=labeled_spans,
+        base_sequence_length=base_sequence_length,
+        offset=offset,
+        include_ill_formed=include_ill_formed,
+    )
+    if include_ill_formed:
+        tags = fix_encoding(tags, "IOB2")
+    else:
+        tags = remove_encoding(tags, "IOB2")
+
+    # Recode the labels if necessary.
+    if coding_scheme == "BIOUL":
+        coded_tags = _to_bioul(tags, encoding="IOB2") if tags is not None else None
+    elif coding_scheme == "BOUL":
+        coded_tags = _to_boul(tags, encoding="IOB2") if tags is not None else None
+    elif coding_scheme == "IOB2":
+        coded_tags = tags
+    else:
+        raise ValueError(f"Unknown Coding scheme {coding_scheme}.")
+
+    return coded_tags
+
+
+def tag_sequence_to_token_spans(
+    tag_sequence: List[str],
+    coding_scheme: str = "IOB2",
+    classes_to_ignore: List[str] = None,
+    include_ill_formed: bool = True,
+):
+
+    if include_ill_formed:
+        new_tag_sequence = fix_encoding(tag_sequence, coding_scheme)
+    else:
+        new_tag_sequence = remove_encoding(tag_sequence, coding_scheme)
+    if coding_scheme == "BIOUL":
+        labeled_spans = bioul_tags_to_spans(
+            new_tag_sequence,
+            classes_to_ignore=classes_to_ignore,
+        )
+    elif coding_scheme == "BOUL":
+        labeled_spans = boul_tags_to_spans(
+            new_tag_sequence,
+            classes_to_ignore=classes_to_ignore,
+        )
+    else:  # coding_scheme = "IOB2":
+        labeled_spans = iob2_tags_to_spans(
+            tag_sequence,
+            classes_to_ignore=classes_to_ignore,
+        )
+        # handled in fix or remove encoding
+    """else:
+        raise ValueError(f"Unknown Coding scheme {coding_scheme}.")"""
+
+    return labeled_spans
+
+
+def _to_boul(tag_sequence: List[str], encoding: str = "IOB2") -> List[str]:
```

Comment:
does `encoding` make any difference? remove?  if so, also rename the method

Reply:
Can we also rename all the methods that expect a specific input encoding accordingly e.g. `_iob2_to_boul` for this case?